### ISSUE TITLE
CFP: Writing Cilium logs to a socket/file in addition to stdout

### ISSUE THREAD

**Issue by tamilmani1989** [cilium/cilium#34908]
## Cilium Feature Proposal

**Describe the feature you'd like**

Currently, the Cilium Agent writes its logs to stdout, and containerd eventually writes these logs to a folder under `/var/log/pods`. In AKS, we are considering exporting Cilium Agent logs to our telemetry infrastructure. To achieve this, we may need to read the logs output by the Cilium Agent to stdout. One option is to have a logging service and mount /var/log/pods, but this would mount all pod logs on the node, which is undesirable as it includes customer pods too. Can Cilium Agent write logs to a file or socket in addition to stdout? This could be an opt-in feature and not enabled by default, allowing our logging service to export logs to the telemetry backend.

We did consider redirecting stdout logs to a file but `kubectl logs` may not work. Wanted to check if this something we can support in cilium agent itself without work around.

Wanted to get feedback before start delving into details of how this can be designed.

**Comment by joestringer** [comment]
Summarizing my perspective from the discussion during today's community meeting:
- The use case makes sense and should be achievable one way or another.
- One question is whether we can achieve this with off-the-shelf components, existing slog library etc. or whether this involves writing a bunch of custom logging code in Cilium. We're not the only application with these requirements, so if we can leverage known similar patterns then that can provide more consistency for the user.
- Another question is whether you intend for logs to be emitted in two places. Logging can have a cost and if we can integrate it in a way that avoids duplicate effort, that would be a nice property to have. 
- If Cilium is outputting these logs to stdout, then is this a question for how the container runtime manages the logs? Is it Cilium's responsibility or not? I'm not sure I have a strong opinion about this yet.
- If this is about an integration with a particular other system, does it make sense to just make sure that the files are logged to a known specific path, then there's another dedicated component which is responsible to slurp up the logs and forward them to the log aggregator?
- If we end up moving towards a solution where Cilium logs to files, then it will be important to handle log rotation to avoid filling up the disk with logs. We've used `lumberjack` for this purpose in the past. I see there's a logging hook for this already. Maybe that pointer will help you to dig around to come up with a more concrete proposal.
- Cilium currently has about 3-4 different loggers - primarily logrus, though we are starting to shift logrus to the standard Go slog implementation instead (example: https://github.com/cilium/cilium/pull/32469), and then there's klog from k8s libraries. Oh and sometimes Cilium will also forward Envoy logs through to stdout. So however we implement this would ideally capture all the logs from different logging systems.

**Comment by tamilmani1989** [comment]
You are right that not all packages are converted to slog and there are still packages using different loggers. If we want to add hook support now, then we may need to support for all loggers which I feel would be a stretch. Also, slog doesn't support adding multiple sinks out of the box, so we may need to build something custom to log to different outputs. I would also like to evaluate if there any overhead/complications sending logs over unix domain socket instead of file logging.

> If Cilium is outputting these logs to stdout, then is this a question for how the container runtime manages the logs? Is it Cilium's responsibility or not? I'm not sure I have a strong opinion about this yet.

containerd redirect stdout logs to specific directory (`/var/log/pods`) under node but we don't want logging agent to mount `/var/log/pods` as it includes other pod logs too. Not sure if we can add custom plugin to containerd to redirect only cilium logs to different location.

> Another question is whether you intend for logs to be emitted in two places. Logging can have a cost and if we can integrate it in a way that avoids duplicate effort, that would be a nice property to have.

Yes both places and it's opt-in. By default, it logs only to stdout. i see there is an option already exposed in cilium config for writing to syslog. https://github.com/cilium/cilium/blob/3049ac6fdef4f8397c41c466a0f1df798cbe5b87/pkg/logging/logging.go#L197

**Comment by aanm** [comment]
Reopened as discussed in the weekly community meeting
